fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,663 @@
+	maxX      float64
+	minX      float64
+	maxY      float64
+	minY      float64
+}
+
+// creates a delaunay triangulation with the given points
+// all points have to be inside the user defined bounds
+func NewDelaunay(points []*Point, maxX float64, minX float64, maxY float64, minY float64) *Delaunay {
+	// root Triangle is a triangle that contains all points
+	dx := maxX - minX
+	dy := maxY - minY
+	a := NewPoint(minX-8*dx, maxY+10, -1)
+	b := NewPoint(maxX+8*dx, maxY+10, -1)
+	c := NewPoint(minX+dx/2, minY-8*dy, -1)
+	root := NewTriangle(a, b, c)
+	d := &Delaunay{
+		root: root,
+		maxX: maxX,
+		minX: minX,
+		maxY: maxY,
+		minY: minY,
+	}
+	d.triangles = make([]*Triangle, 0)
+	for _, p := range points {
+		d.InsertPoint(p)
+	}
+	return d
+}
+
+func (d *Delaunay) GetTriangles() []*Triangle {
+	// remove triangles that contain the root points
+	rt := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+	rt = append(rt, d.root.A.adjacentTriangles[:]...)
+	rt = append(rt, d.root.B.adjacentTriangles[:]...)
+	rt = append(rt, d.root.C.adjacentTriangles[:]...)
+	return finalize(d.triangles, rt[:]...)
+}
+
+func (d *Delaunay) InsertPoint(p *Point) {
+	p.adjacentTriangles = make([]*Triangle, 0)
+	t, onE := findTriangle(d.root, p)
+	if t == nil {
+		// should only happen when user gives wrong max and min values
+		panic(fmt.Errorf("delaunay: no triangle which contains P%s", p))
+	}
+	if onE {
+		d.insertPonE(t, p)
+	} else {
+		d.insertP(t, p)
+	}
+}
+
+func (d *Delaunay) RemovePoint(p *Point) {
+	// if that's the case either an error has occurred or the point is outside or on the root triangle
+	if len(p.adjacentTriangles) < 3 {
+		// should only happen when user makes a mistake
+		panic(fmt.Errorf("delaunay: can't remove point P%s", p))
+	}
+	// remove triangles adjacent to p
+	for _, t := range p.adjacentTriangles {
+		if t.A.Equals(p) {
+			t.B.adjacentTriangles = remove(t.B.adjacentTriangles, t)
+			t.C.adjacentTriangles = remove(t.C.adjacentTriangles, t)
+		} else if t.B.Equals(p) {
+			t.A.adjacentTriangles = remove(t.A.adjacentTriangles, t)
+			t.C.adjacentTriangles = remove(t.C.adjacentTriangles, t)
+		} else if t.C.Equals(p) {
+			t.A.adjacentTriangles = remove(t.A.adjacentTriangles, t)
+			t.B.adjacentTriangles = remove(t.B.adjacentTriangles, t)
+		} else {
+			// should never happen
+			panic(fmt.Errorf("delaunay: internal error with adjacent triangles for P%s and T%s", p, t))
+		}
+	}
+	// find points on polygon around the point
+	points := make([]*Point, len(p.adjacentTriangles))
+	t := p.adjacentTriangles[0]
+	j := 1
+	k := 0
+	if t.A.Equals(p) {
+		points[0] = t.B
+		points[1] = t.C
+	} else if t.B.Equals(p) {
+		points[0] = t.C
+		points[1] = t.A
+	} else if t.C.Equals(p) {
+		points[0] = t.A
+		points[1] = t.B
+	} else {
+		// should never happen
+		panic(fmt.Errorf("delaunay: internal error with adjacent triangles for P%s and T%s", p, t))
```

Comment:
`panic(fmt.Errorf("delaunay: point %v not in adjacent triangle %v", p, t))`

BTW, this could just be fmt.Sprintf, unless you are wanting people to recover and type switch on the recovered value. Not a big issue, but something to consider in the future.